applications/IgaApplication/custom_elements/surface_base_discrete_element.cpp

```diff
@@ -0,0 +1,762 @@
+        //const int mat_size = number_of_control_points * 3;
+
+        //for (int r = 0; r < mat_size; r++)
+        //{
+        //    Vector zu = ZeroVector(3);
+
+        //        // dN_ca = Dm*S_dE_ca(:,r);
+        //    zu[0] = D(0, 0)*B(0, r) + D(0, 1)*B(1, r) + D(0, 2)*B(2, r);
+        //    zu[1] = D(1, 0)*B(0, r) + D(1, 1)*B(1, r) + D(1, 2)*B(2, r);
+        //    zu[2] = D(2, 0)*B(0, r) + D(2, 1)*B(1, r) + D(2, 2)*B(2, r);
+
+
+        //    for (int s = 0; s <= r; s++)
+        //    {
+        //        // membrane stiffness
+        //        double x = (zu[0] * B(0, s) + zu[1] * B(1, s) + zu[2] * B(2, s))*IntegrationWeight;
+        //        rLeftHandSideMatrix(r, s) += x;
+        //        rLeftHandSideMatrix(s, r) += x;
+        //        //// bending stiffness
+        //        //S_keb(r, s) = dM_ca[0] * S_dK_ca(0, s) + dM_ca[1] * S_dK_ca(1, s) + dM_ca[2] * S_dK_ca(2, s)
+        //        //    + M_ca[0] * S_ddK_ca_1(r, s) + M_ca[1] * S_ddK_ca_2(r, s) + M_ca[2] * S_ddK_ca_3(r, s);
+        //        //S_keb(s, r) = S_keb(r, s);
+        //    }
+        //}
+        KRATOS_CATCH("")
+    }
+
+    //***********************************************************************************
+    //***********************************************************************************
+    void SurfaceBaseDiscreteElement::CalculateAndAddNonlinearKm(
+        Matrix& rLeftHandSideMatrix,
+        const SecondVariations& SecondVariationsStrain,
+        const Vector& SD,
+        const double& rIntegrationWeight)
+
+    {
+        KRATOS_TRY
+        const int number_of_control_points = GetGeometry().size();
+        const int mat_size = number_of_control_points * 3;
+
+        for (int n = 0; n < mat_size; n++)
+        {
+            for (unsigned int m = 0; m <= n; m++)
+            {
+                double nm = (SD[0] * SecondVariationsStrain.B11(n, m)
+                    + SD[1] * SecondVariationsStrain.B22(n, m)
+                    + SD[2] * SecondVariationsStrain.B12(n, m))*rIntegrationWeight;
+                //KRATOS_ERROR_IF(nm > 1e-5) << "Problem" << std::endl;
+                rLeftHandSideMatrix(n, m) += nm;
+                rLeftHandSideMatrix(m, n) += nm;
+            }
+        }
+        KRATOS_CATCH("")
+    }
+    //************************************************************************************
+    //************************************************************************************
+    void SurfaceBaseDiscreteElement::CalculateMetric(
+        MetricVariables& metric)
+    {
+        KRATOS_ERROR << "You have called to the CalculateMetric from the base class for meshless surface elements" << std::endl;
+    }
+
+    //***********************************************************************************
+    //***********************************************************************************
+    void SurfaceBaseDiscreteElement::CalculateStrain(
```

Comment:
Could be moved out of the class or made `static` since it is independent of the object.